skcosmo/selection/FPS.py

```diff
@@ -0,0 +1,171 @@
+            self.idx = idxs
+        else:
+            self.idx = [np.random.randint(self.product.shape[0])]
+        self.distances = np.min(
+            [_calculate_distances_(self.product, idx) for idx in self.idx], axis=0
+        )
+        self.distance_selected = np.nan * np.zeros(self.product.shape[0])
+
+    def select(self, n):
+        """
+        Method for FPS select based upon a product of the input matrices
+
+        :param n: number of selections to make
+        :type n: int
+
+        :returns: list of selected indices
+        :rtype: list of int
+        """
+        if len(self.idx) > n:
+            return self.idx[:n]
+
+        # Loop over the remaining points...
+        for i in range(len(self.idx) - 1, n - 1):
+            for j in np.where(self.distances > 0)[0]:
+                self.distances[j] = min(
+                    self.distances[j],
+                    _calculate_distances_(self.product, self.idx[i], [j]),
+                )
+            self.idx.append(np.argmax(self.distances))
+            self.distance_selected[self.idx[-1]] = self.distances[self.idx[-1]]
+
+            if np.abs(self.distances).max() < self.tol:
+                return self.idx
+        return self.idx
+
+
+class sampleFPS(_FPS):
+    """
+
+    For sample selection, traditional FPS employs a row-wise Euclidean
+    distance, which can be expressed using the Gram matrix
+    :math:`\\mathbf{K} = \\mathbf{X} \\mathbf{X}^T`
+
+    .. math::
+        \\operatorname{d}_r(i, j) = K_{ii} - 2 K_{ij} + K_{jj}.
+
+    When mixing < 1, this will use PCov-FPS, where a modified Gram matrix is
+    used to express the distances
+
+    .. math::
+        \\mathbf{\\tilde{K}} = \\alpha \\mathbf{XX}^T +
+        (1 - \\alpha)\\mathbf{\\hat{Y}\\hat{Y}}^T
+
+    :param idxs: predetermined indices; if None provided, first index selected
+                 is random
+    :type idxs: list of int, None
+
+    :param matrix: Data matrix :math:`\\mathbf{X}` from which to select a
+                   subset of the `n` rows
+    :type matrix: array of shape (n x m)
+
+    :param mixing: mixing parameter, as described in PCovR as
+                  :math:`{\\alpha}`, defaults to 1
+    :type mixing: float
+
+    :param tolerance: threshold below which values will be considered 0,
+                      defaults to 1E-12
+    :type tolerance: float
+
+    :param Y: array to include in biased selection when mixing < 1;
+              required when mixing < 1, throws AssertionError otherwise
+    :type Y: array of shape (n x p), optional
+
+    """
+
+    def __init__(self, matrix, mixing=1.0, tolerance=1e-12, **kwargs):
+
+        self.mixing = mixing
+        self.tol = tolerance
+
+        self.A = matrix.copy()
+
+        if mixing < 1:
+            try:
+                assert "Y" in kwargs
+                self.Y = kwargs.get("Y")
+            except AssertionError:
+                print(r"For $\mixing < 1$, $Y$ must be in" "the constructor parameters")
```

Comment:
This should be an error, right? There is nothing we can do to save it here

```suggestion
                raise Exception(r"For $\mixing < 1$, $Y$ must be in" "the constructor parameters")
```